### ISSUE TITLE
"No valid adapter found"

### ISSUE THREAD

**Issue by gbell12** [welaika/wordmove#96]
Hi Guys - thanks for this neat tool - bringing command-line power to the somewhat backwards world of PHP CMS's!  Not getting it to work yet though...

On a wordmove push, I get:
```
▬▬ ✓ Using Movefile: ./Movefile ▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬
/var/lib/gems/1.9.1/gems/wordmove-1.2.0/lib/wordmove/deployer/base.rb:36:in `deployer_for': No valid adapter found. (StandardError)
        from /var/lib/gems/1.9.1/gems/wordmove-1.2.0/lib/wordmove/cli.rb:56:in `push'
        from /var/lib/gems/1.9.1/gems/thor-0.19.1/lib/thor/command.rb:27:in `run'
        from /var/lib/gems/1.9.1/gems/thor-0.19.1/lib/thor/invocation.rb:126:in `invoke_command'
        from /var/lib/gems/1.9.1/gems/thor-0.19.1/lib/thor.rb:359:in `dispatch'
        from /var/lib/gems/1.9.1/gems/thor-0.19.1/lib/thor/base.rb:440:in `start'
        from /var/lib/gems/1.9.1/gems/wordmove-1.2.0/bin/wordmove:6:in `<top (required)>'
        from /usr/local/bin/wordmove:23:in `load'
        from /usr/local/bin/wordmove:23:in `<main>'
```

**Comment by delphaber** [comment]
@gbell12 any news? :)

**Comment by gbell12** [comment]
I don't believe anybody's waiting on me for anything...

**Comment by pioneerskies** [comment]
@gbell12 I think @delphaber "pinged" you because the troubleshooting process was interrupted w/o a good ending. We would be pleased to continue helping you or eventually close the ticket if no help is needed anymore.

**Comment by gbell12** [comment]
I edited all my comments to properly enclose code as per one of the comments above.

Last comment was "I've looked into the code and i can just confirm that Wordmove isn't reading your movefile's ssh section for some arcane reason"

Not sure what else I need to do.  A stack dump is obviously not a great thing for the user to see, but conveys a lot of info to the devs, so I'm thinking that's all the devs needed.

**Comment by pioneerskies** [comment]
Ok. The only un-replied question was actually `Are you running the wordmove command from the correct folder?`. We will assume "yes" when doing the next debugging rush on the software.

Just as a memo to developers: another, non noticed by the user, difference between original YAML and my reformatted YAML was

```
ssh:
host: "bff.com.au"
user: "bff"
```

vs

```
ssh:
  host: "bff.com.au"
  user: "bff"
```

Would be good to try to reproduce the error w/ that configuration.

Cheers

**Comment by gbell12** [comment]
Because of the absolute paths, I wasn't aware that it needed to be run from a particular place.
If I indent my ssh: section, it runs, instantaneously giving this output:

```
▬▬ ✓ Using Movefile: ./Movefile ▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬
```
Which I don't think is actually doing anything unless my internet connection has been upgraded to "instantaneous" :)

**Comment by pioneerskies** [comment]
Hey @gbell12 overall I'm glad we went some steps further +1

About pretty confusing messages, we are aware of it as per #127 and we'll work on that.

About your output I think you are in a similar situation to the aforementioned issue:

```
↳ $ wordmove pull

▬▬ ✓ Using Movefile: ./Movefile ▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬
```

you have to invoke `pull`/`push` commands followed by one of the flag from `wordmove help push`, e.g.:

```
↳ $ wordmove pull -t

▬▬ ✓ Using Movefile: ./Movefile ▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬

▬▬ ✓ Pulling Themes ▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬
   remote | get_directory: /home/163339/domains/welaika.demo.welaika.com/html/wp-content/themes /Users/fuzzy/dev/xxxxxxx/wp-content/themes .git/ .gitignore .sass-cache/ bin/ tmp/* Gemfile* Movefile wp-config.php wp-content/*.sql
I, [2015-03-24T11:15:46.405562 #3604]  INFO -- : rsync --progress -e 'ssh -p 22 xxxxxx@example.com' -rlpt --compress --omit-dir-times --delete --exclude .git/ --exclude .gitignore --exclude .sass-cache/ --exclude bin/ --exclude 'tmp/*' --exclude 'Gemfile*' --exclude Movefile --exclude wp-config.php --exclude 'wp-content/*.sql' :/xxxxxxxx/wp-content/themes/ /Users/fuzzy/dev/welaika.com-2015/wp-content/themes
receiving file list ...
113 files to consider
./
.DS_Store
        6148 100%    5.86MB/s    0:00:00 (xfer#1, to-check=111/113)
[cut]
```

If I am not on the right way we would consider to open a new issue. Let me know

Cheers and thanks for the follow-up

**Comment by taichi2000** [comment]
Yes, for me that was the issue: YAML Indentation. I put ssh config entry as at the same level for staging node, but it should be as a child of it. Oh gosh. It is working fine now.

**Comment by gbell12** [comment]
Working fine now.  Needed a -t.  Thanks for the support and the awesome tool!

**Comment by charmcat** [comment]
Okay, so I'm having the same error. Not sure if I should open a new thread or not? What's the etiquette??

**Comment by pioneerskies** [comment]
Hey you can try to paste here your `Movefile`. Please keep it pre-formatted while pasting. I'm quite sure this error wes, is and will be always due to wrong indentation in `Movefile`.

We're thinking about a built-in validator, but actually not having enough time do implement :/

**Comment by charmcat** [comment]
````
dev:
  vhost: "http://site.dev"
  wordpress_path: "/Users/me/Sites/site-local"
 
  database:
    name: "wordpress"
    user: "root"
    password: "root"
    host: "127.0.0.1"
 
test:
  vhost: "http://site.net"
  wordpress_path: "/dev/www/"
 
  database:
    name: "me"
    user: "me"
    password: "pass"
    host: "host"
 
  exclude:
    - ".git/"
    - ".gitignore"
    - ".sass-cache/"
    - "bin/"
    - "tmp/*"
    - "Gemfile*"
    - "Movefile"
    - "wp-config.php"
    - "wp-content/*.sql"
    
  ssh:
    host: "host"
    user: "me"
    password: "pass"
 
live:
  vhost: "http://site.net"
  wordpress_path: "/site/www/"
 
  database:
    name: "me"
    user: "me"
    password: "pass"
    host: "host"
 
  exclude:
    - ".git/"
    - ".gitignore"
    - ".sass-cache/"
    - "bin/"
    - "tmp/*"
    - "Gemfile*"
    - "Movefile"
    - "wp-config.php"
    - "wp-content/*.sql"
 
  ssh:
    host: "host"
    user: "me"
    password: "pass"
````

It validates okay.

**Comment by pioneerskies** [comment]
@charmcat could you also write down the command you used? Sorry if I asked just now :(

**Comment by charmcat** [comment]
````
wordmove pull -e dev --all
````

**Comment by pioneerskies** [comment]
OK @charmcat so we have just a misunderstanding as far as I understand :)

`-e` argument takes **always** remote environments. `Wordmove` assumes one of the envs will always be your development env. So:

1. Never modify the first part's name of the `Movefile` `local:`. That is you local dev env. Full stop
2. When you pull/push you can specify a remote w/ `-e` flag if you have multiple remotes (the environments other than the `local` one). If you have just one remote it will be the default w/o `-e`
3. When you pull the destination will always be `local`; tell `Wordmove` the environment from where it should take the bits w/ `-e`
4. When you push the destination will always be the requested remote; tell `Wordmove` the required destination environment w/ `-e`

Practically:

```yaml
local:
  vhost: "http://site.dev"
  wordpress_path: "/Users/me/Sites/site-local"

  database:
    name: "wordpress"
    user: "root"
    password: "root"
    host: "127.0.0.1"

test:
  vhost: "http://site.net"
  wordpress_path: "/dev/www/"

  database:
    name: "me"
    user: "me"
    password: "pass"
    host: "host"

  exclude:
    - ".git/"
    - ".gitignore"
    - ".sass-cache/"
    - "bin/"
    - "tmp/*"
    - "Gemfile*"
    - "Movefile"
    - "wp-config.php"
    - "wp-content/*.sql"

  ssh:
    host: "host"
    user: "me"
    password: "pass"

live:
  vhost: "http://site.net"
  wordpress_path: "/site/www/"

  database:
    name: "me"
    user: "me"
    password: "pass"
    host: "host"

  exclude:
    - ".git/"
    - ".gitignore"
    - ".sass-cache/"
    - "bin/"
    - "tmp/*"
    - "Gemfile*"
    - "Movefile"
    - "wp-config.php"
    - "wp-content/*.sql"

  ssh:
    host: "host"
    user: "me"
    password: "pass"
```

Then pull from the live w/

    wordmove pull --all -e live

or pull from the test w/

    wordmove pull --all -e test

I really hope this will explain the basic `Wordmove`'s mechanics.

Please let us know about your progress

**Comment by charmcat** [comment]
lol thanks. I really have no idea about this stuff, so I'm completely in the dark. I used the right command and got a couple of other error messages that I was able to fix, and it seems to be working now!

The environment explanation would be great in the wiki!

**Comment by pioneerskies** [comment]
Here we are @charmcat :) https://github.com/welaika/wordmove/wiki/Multiple-environments-explained

I'm going to re-close this issue.

Cheers
